File: labs/04-offsets-replay/consumer.js
import { decodeMessage, kafka, topicName } from "../../lib/kafka.js";

const topic = topicName(process.env.TOPIC_SUFFIX ?? "orders");
const groupId = process.env.GROUP_ID ?? "replay-demo-v1";
const consumer = kafka.consumer({ groupId });

process.on("SIGINT", async () => {
  await consumer.disconnect();
  process.exit(0);
});

await consumer.connect();
await consumer.subscribe({ topic, fromBeginning: true });

console.log(`Reading ${topic} as group "${groupId}". Press Ctrl+C to stop.`);

await consumer.run({
  eachMessage: async ({ partition, message }) => {
    const event = decodeMessage(message);

    console.log({
      partition,
      offset: message.offset,
      key: event.key,
      value: event.value
    });
  }
});

File: labs/04-offsets-replay/consumer.py
import os
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[2]))

from lab_kafka.client import consumer, decode_message, topic_name


def main() -> None:
    topic = topic_name(os.getenv("TOPIC_SUFFIX", "orders"))
    group_id = os.getenv("GROUP_ID", "replay-demo-v1")
    kafka_consumer = consumer(group_id)
    kafka_consumer.subscribe([topic])

    print(f'Reading {topic} as group "{group_id}". Press Ctrl+C to stop.')

    try:
        while True:
            message = kafka_consumer.poll(1.0)

            if message is None:
                continue

            if message.error():
                print(message.error())
                continue

            event = decode_message(message)
            print(
                {
                    "partition": message.partition(),
                    "offset": message.offset(),
                    "key": event["key"],
                    "value": event["value"],
                }
            )
            kafka_consumer.commit(message=message, asynchronous=False)
    except KeyboardInterrupt:
        print("Stopping consumer.")
    finally:
        kafka_consumer.close()


if __name__ == "__main__":
    main()
